From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wayne Chang <waynec@nvidia.com>
Date: Thu, 14 Nov 2024 03:07:37 +0000
Subject: [PATCH 160/240] NVIDIA: SAUCE: phy: tegra: xusb-tegra186: Fix and
 extend fuse calibration handling

The handling of the HS_TERM_RANGE_ADJ calibration was previously incorrect
due to its differing treatment across SoC generations. On Tegra186 and
earlier, the HS_TERM_RANGE_ADJ value was shared among all UTMI pads.
However, starting with Tegra194 and later, the calibration value is unique
for each pad. This patch fixes the issue by adjusting the HS_TERM_RANGE_ADJ
handling to correctly support per-pad calibration on newer SoCs while
maintaining compatibility with older ones.

Additionally, this patch adds support for REC_OS and SQ_OS calibration on
Tegra264. These calibration parameters, which are relevant for USB performance,
are now read from the appropriate fuse registers and applied to each pad
individually.

Bug 4933929

Change-Id: If449ab6c56f59b73f8015c8d33480c271f1a9518
Signed-off-by: Wayne Chang <waynec@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3248456
Reviewed-by: Henry Lin <henryl@nvidia.com>
Reviewed-by: HaoTien Hsu <haotienh@nvidia.com>
Reviewed-by: WK Tsai <wtsai@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Sing-Han Chen <singhanc@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
(cherry picked from commit 1f219c8da2b127952d506fcc3a9e3b420bfe1962)

diff --git a/drivers/phy/tegra/xusb-tegra186.c b/drivers/phy/tegra/xusb-tegra186.c
index 2f2ad4d7a45d..1b6e6955d942 100644
--- a/drivers/phy/tegra/xusb-tegra186.c
+++ b/drivers/phy/tegra/xusb-tegra186.c
@@ -31,6 +31,10 @@
 
 #define RPD_CTRL_SHIFT			0
 #define RPD_CTRL_MASK			0x1f
+#define REC_OS_PADX_SHIFT(x)	        (17 + (x) * 4)
+#define REC_OS_PAD_MASK		        0xf
+#define SQ_OS_PADX_SHIFT(x)	        ((x) * 8)
+#define SQ_OS_PAD_MASK		        0xff
 
 /* XUSB PADCTL registers */
 #define XUSB_PADCTL_USB2_PAD_MUX	0x4
@@ -85,10 +89,13 @@
 #define  TERM_RANGE_ADJ(x)			(((x) & 0xf) << 3)
 #define  RPD_CTRL(x)				(((x) & 0x1f) << 26)
 
+#define XUSB_PADCTL_USB2_OTG_PADX_CTL3(x)	(0x94 + (x) * 0x40)
+#define  REC_OS(x)			        (((x) & 0xf) << 9)
+#define  SQ_OS(x)				(((x) & 0xff) << 14)
+
 #define XUSB_PADCTL_USB2_BIAS_PAD_CTL0		0x284
 #define  BIAS_PAD_PD				BIT(11)
 #define  HS_SQUELCH_LEVEL(x)			(((x) & 0x7) << 0)
-#define  HS_DISCON_LEVEL(x)			(((x) & 0x7) << 3)
 
 #define XUSB_PADCTL_USB2_BIAS_PAD_CTL1		0x288
 #define  USB2_TRK_START_TIMER(x)		(((x) & 0x7f) << 12)
@@ -260,8 +267,6 @@
 #define USB2_BIAS_PAD_CTL1_TRK_DONE_RESET_TIMER_FIELD_START	19
 #define USB2_BIAS_PAD_CTL1_TRK_DONE_RESET_TIMER_FIELD_WIDTH	7
 
-#define XUSB_PADCTL_USB2_OTG_PADX_CTL3(x)		(0x94 + (x) * 0x40)
-
 #define TEGRA186_LANE(_name, _offset, _shift, _mask, _type)		\
 	{								\
 		.name = _name,						\
@@ -277,6 +282,8 @@ struct tegra_xusb_fuse_calibration {
 	u32 hs_squelch;
 	u32 *hs_term_range_adj;
 	u32 rpd_ctrl;
+	u32 *sq_os;
+	u32 *rec_os;
 };
 
 struct tegra186_xusb_padctl_context {
@@ -652,8 +659,6 @@ static void tegra186_utmi_bias_pad_power_on(struct tegra_xusb_padctl *padctl)
 	value &= ~BIAS_PAD_PD;
 	value &= ~HS_SQUELCH_LEVEL(~0);
 	value |= HS_SQUELCH_LEVEL(priv->calib.hs_squelch);
-	value &= ~HS_DISCON_LEVEL(~0);
-	value |= HS_DISCON_LEVEL(0x7);
 	padctl_writel(padctl, value, XUSB_PADCTL_USB2_BIAS_PAD_CTL0);
 
 	udelay(1);
@@ -989,6 +994,15 @@ static int tegra186_utmi_phy_power_on(struct phy *phy)
 	value |= RPD_CTRL(priv->calib.rpd_ctrl);
 	padctl_writel(padctl, value, XUSB_PADCTL_USB2_OTG_PADX_CTL1(index));
 
+	if (priv->calib.rec_os && priv->calib.sq_os) {
+		value = padctl_readl(padctl, XUSB_PADCTL_USB2_OTG_PADX_CTL3(index));
+		value &= ~REC_OS(~0);
+		value |= REC_OS(priv->calib.rec_os[index]);
+		value &= ~SQ_OS(~0);
+		value |= SQ_OS(priv->calib.sq_os[index]);
+		padctl_writel(padctl, value, XUSB_PADCTL_USB2_OTG_PADX_CTL3(index));
+	}
+
 skip_fuse_calibration:
 	tegra186_utmi_pad_power_on(phy);
 
@@ -1521,14 +1535,13 @@ static const char * const tegra186_usb3_functions[] = {
 static int
 tegra186_xusb_read_fuse_calibration(struct tegra186_xusb_padctl *padctl)
 {
-	const struct tegra_xusb_padctl_soc *soc = padctl->base.soc;
 	struct device *dev = padctl->base.dev;
 	unsigned int i, count;
-	u32 value, *level;
-	u32 *hs_term_range_adj;
+	u32 value, *level, *hs_term_range_adj;
 	int err;
+	u32 chip_id = tegra_get_chip_id();
 
-	count = soc->ports.usb2.count;
+	count = padctl->base.soc->ports.usb2.count;
 
 	level = devm_kcalloc(dev, count, sizeof(u32), GFP_KERNEL);
 	if (!level)
@@ -1553,30 +1566,66 @@ tegra186_xusb_read_fuse_calibration(struct tegra186_xusb_padctl *padctl)
 
 	padctl->calib.hs_squelch = (value >> HS_SQUELCH_SHIFT) &
 					HS_SQUELCH_MASK;
+
 	hs_term_range_adj[0] = (value >> HS_TERM_RANGE_ADJ_PADX_SHIFT(0)) &
-				HS_TERM_RANGE_ADJ_PAD_MASK;
+					HS_TERM_RANGE_ADJ_PAD_MASK;
 
 	err = tegra_fuse_readl(TEGRA_FUSE_USB_CALIB_EXT_0, &value);
-	if (err) {
-		dev_err(dev, "failed to read calibration fuse: %d\n", err);
-		return err;
-	}
+	if (err)
+		return dev_err_probe(dev, err,
+				     "failed to read EXT0 calibration fuse\n");
 
 	dev_dbg(dev, "FUSE_USB_CALIB_EXT_0 %#x\n", value);
 
 	padctl->calib.rpd_ctrl = (value >> RPD_CTRL_SHIFT) & RPD_CTRL_MASK;
 
 	for (i = 1; i < count; i++) {
-		if (soc->has_per_pad_term)
-			hs_term_range_adj[i] =
-				(value >> HS_TERM_RANGE_ADJ_PADX_SHIFT(i)) &
-				HS_TERM_RANGE_ADJ_PAD_MASK;
+		if (chip_id && chip_id > TEGRA186)
+			hs_term_range_adj[i] = (value >> HS_TERM_RANGE_ADJ_PADX_SHIFT(i)) &
+						    HS_TERM_RANGE_ADJ_PAD_MASK;
 		else
 			hs_term_range_adj[i] = hs_term_range_adj[0];
 	}
 
 	padctl->calib.hs_term_range_adj = hs_term_range_adj;
 
+	if (chip_id && chip_id > TEGRA234) {
+		u32 *sq_os, *rec_os;
+
+		sq_os = devm_kcalloc(dev, count, sizeof(u32), GFP_KERNEL);
+		if (!sq_os)
+			return -ENOMEM;
+
+		rec_os = devm_kcalloc(dev, count, sizeof(u32), GFP_KERNEL);
+		if (!rec_os)
+			return -ENOMEM;
+
+		for (i = 0; i < 3; i++)
+			rec_os[i] = (value >> REC_OS_PADX_SHIFT(i)) & REC_OS_PAD_MASK;
+
+		err = tegra_fuse_readl(TEGRA_FUSE_USB_CALIB_EXT3_0, &value);
+		if (err)
+			return dev_err_probe(dev, err,
+					     "failed to read EXT3 calibration fuse\n");
+
+		dev_dbg(dev, "FUSE_USB_CALIB_EXT3_0 %#x\n", value);
+
+		rec_os[3] = value & REC_OS_PAD_MASK;
+
+		err = tegra_fuse_readl(TEGRA_FUSE_USB_CALIB_EXT2_0, &value);
+		if (err)
+			return dev_err_probe(dev, err,
+					     "failed to read EXT2 calibration fuse\n");
+
+		dev_dbg(dev, "FUSE_USB_CALIB_EXT2_0 %#x\n", value);
+
+		for (i = 0; i < count; i++)
+			sq_os[i] = (value >> SQ_OS_PADX_SHIFT(i)) & SQ_OS_PAD_MASK;
+
+		padctl->calib.rec_os = rec_os;
+		padctl->calib.sq_os = sq_os;
+	}
+
 	return 0;
 }
 
@@ -1795,7 +1844,6 @@ const struct tegra_xusb_padctl_soc tegra194_xusb_padctl_soc = {
 	.num_supplies = ARRAY_SIZE(tegra194_xusb_padctl_supply_names),
 	.supports_gen2 = true,
 	.poll_trk_completed = true,
-	.has_per_pad_term = true,
 };
 EXPORT_SYMBOL_GPL(tegra194_xusb_padctl_soc);
 
@@ -1820,7 +1868,6 @@ const struct tegra_xusb_padctl_soc tegra234_xusb_padctl_soc = {
 	.trk_hw_mode = false,
 	.trk_update_on_idle = true,
 	.supports_lp_cfg_en = true,
-	.has_per_pad_term = true,
 };
 EXPORT_SYMBOL_GPL(tegra234_xusb_padctl_soc);
 
-- 
2.55.0

